README.md
# Break Time README

This extension will remind to take a break during your coding sessions.

## Features

### **_Starting_**

-Just click the **next break button** in the status bar.

![](img/NextBreak.png)

-A **quick pick** will show up, pick the time and a timer will be set.

![](img/QuickPick.png)

-In the status bar the extension will tell you the **time left** for break time.
![](img/BreakIn.png)

-When is **break time** an information **message** will pop up.

![](img/BreakMessage.png)

-If you press **another 5 minutes** a new timer will be set to 5 mins.

### **_Stop the timer_**

-To **stop** the timer, press the button and click STOP at the top of the quick pick.

![](img/StopBreak.png)

## Extension Settings

This extension contributes the following settings:

-   `BreakTime.enable`: enable/disable this extension.
-   `BreakTime.redText`: Set when the time will be displayed in red.
    ![](img/RedBreak.png)
-   `BreakTime.showLastMinute`: Enable to see seconds last minute. Default will display <1 min.
-   `BreakTime.endMessages`: custom your own messages to display when break time.

          Example:

          "BreakTime.endMessages": ["go take a walk!", "call the boss!", "stretch the leg!"]

## **Enjoy!**


CHANGELOG.md
# Change Log

All notable changes to the "break-time" extension will be documented in this file.

## [1.0.1]

-   New feature: by default last 1 min will be displayed as <1 min. It can be changed in user settings.
-   Minor changes in readme file.

## [1.0.0]

-   Initial release


out/Timer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Timer = void 0;
const vscode = require("vscode");
class Timer {
    constructor(statusBarItem, statusBarTxt) {
        this.config = vscode.workspace.getConfiguration("BreakTime");
        this.redText = 0;
        this.endMessages = ["get up and do some stretching!!", "break time!"];
        this.intervalID = undefined;
        this.currentTime = 0; //seconds
        this.running = false;
        this.statusBarItem = statusBarItem;
        this.statusBarTxt = statusBarTxt;
        if (this.config.redText !== "inactive") {
            this.setRedText();
        }
        this.setEndMessages();
    }
    start(time) {
        if (time === 0) {
            //the user stopped the timer
            this.resetTimer();
            return;
        }
        this.currentTime = time;
        this.running = true;
        if (this.intervalID) {
            clearInterval(this.intervalID);
        }
        //set interval that handle the count down
        this.intervalID = setInterval(() => {
            if (this.currentTime <= 0) {
                //break time!
                this.resetTimer();
                const randomMessage = this.endMessages[Math.floor(Math.random() * this.endMessages.length)];
                vscode.window.showInformationMessage(`${Math.round(time / 60)} mins passed, ${randomMessage}`, "Another 5 minutes").then(() => {
                    this.start(5 * 60);
                });
            }
            else {
                //decrease time and update the text in UI
                this.currentTime--;
                this.statusBarItem.text = `Break in: ${this.secToTime(this.currentTime)}`;
                //last tot sec display red text
                const color = "firebrick";
                if (this.currentTime < this.redText && this.statusBarItem.color !== color) {
                    this.statusBarItem.color = color;
                }
            }
        }, 1000);
    }
    isRunning() {
        return this.running;
    }
    resetTimer() {
        this.statusBarItem.text = this.statusBarTxt;
        clearInterval(this.intervalID);
        this.statusBarItem.color = undefined;
        this.currentTime = 0;
        this.intervalID = undefined;
        this.running = false;
        this.statusBarItem.color = undefined;
    }
    secToTime(totalSecs) {
        const hours = Math.floor(totalSecs / 3600);
        const minutes = Math.floor((totalSecs % 3600) / 60);
        const seconds = Math.floor(totalSecs % 60);
        return `${hours > 0 ? `${hours}h ` : ""}${minutes > 0 ? `${minutes}min ` : ""}${totalSecs <= 60 ? `${seconds}sec` : ""}`;
    }
    setRedText() {
        const red = this.config.redText;
        let num = 0;
        if (red.slice(0, 1) !== "3") {
            num = Number(red.slice(0, 1)) * 60;
        }
        else {
            num = 30;
        }
        this.redText = num;
    }
    setEndMessages() {
        const msg = this.config.endMessages;
        console.log(msg);
        if (msg && Array.isArray(msg)) {
            let valid = true;
            msg.forEach(m => {
                if (typeof m !== "string") {
                    valid = false;
                }
            });
            if (valid === true) {
                this.endMessages = msg;
            }
        }
    }
}
exports.Timer = Timer;
//# sourceMappingURL=Timer.js.map

src/Timer.ts
import * as vscode from "vscode";

export class Timer {
	private config = vscode.workspace.getConfiguration("BreakTime");
	private redText: number = 0;
	private endMessages: string[] = ["get up and do some stretching!!", "break time!"];

	private intervalID: any = undefined;
	private currentTime: number = 0; //seconds
	private running: boolean = false;

	private statusBarItem: vscode.StatusBarItem;
	private statusBarTxt: string;

	constructor(statusBarItem: vscode.StatusBarItem, statusBarTxt: string) {
		this.statusBarItem = statusBarItem;
		this.statusBarTxt = statusBarTxt;
		if (this.config.redText !== "inactive") {
			this.setRedText();
		}
		this.setEndMessages();
	}

	public start(time: number) {
		if (time === 0) {
			//the user stopped the timer
			this.resetTimer();
			return;
		}

		this.currentTime = time;
		this.running = true;

		if (this.intervalID) {
			clearInterval(this.intervalID);
		}

		//set interval that handle the count down
		this.intervalID = setInterval(() => {
			if (this.currentTime <= 0) {
				//break time!
				this.resetTimer();
				const randomMessage = this.endMessages[Math.floor(Math.random() * this.endMessages.length)];
				vscode.window.showInformationMessage(`${Math.round(time / 60)} mins passed, ${randomMessage}`, "Another 5 minutes").then(() => {
					this.start(5 * 60);
				});
			} else {
				//decrease time and update the text in UI
				this.currentTime--;
				//check settings about last minute to display or not
				if (this.currentTime <= 60 && this.config.showLastMinute === false) {
					this.statusBarItem.text = `Break in: <1 min`;
				} else {
					this.statusBarItem.text = `Break in: ${this.secToTime(this.currentTime)}`;
				}

				//last tot sec display red text
				const color: string = "firebrick";
				if (this.currentTime < this.redText && this.statusBarItem.color !== color) {
					this.statusBarItem.color = color;
				}
			}
		}, 1000);
	}

	public isRunning() {
		return this.running;
	}

	private resetTimer() {
		this.statusBarItem.text = this.statusBarTxt;
		clearInterval(this.intervalID);
		this.statusBarItem.color = undefined;
		this.currentTime = 0;
		this.intervalID = undefined;
		this.running = false;
		this.statusBarItem.color = undefined;
	}

	private secToTime(totalSecs: number) {
		const hours = Math.floor(totalSecs / 3600);
		const minutes = Math.floor((totalSecs % 3600) / 60);
		const seconds = Math.floor(totalSecs % 60);
		return `${hours > 0 ? `${hours}h ` : ""}${minutes > 0 ? `${minutes}min ` : ""}${totalSecs <= 60 ? `${seconds}sec` : ""}`;
	}

	private setRedText() {
		const red = this.config.redText;
		let num: number = 0;
		if (red.slice(0, 1) !== "3") {
			num = Number(red.slice(0, 1)) * 60;
		} else {
			num = 30;
		}
		this.redText = num;
	}

	private setEndMessages() {
		const msg = this.config.endMessages;
		if (msg[0] !== undefined && Array.isArray(msg)) {
			//see if every element of the array is a string
			let valid = true;
			msg.forEach(m => {
				if (typeof m !== "string") {
					valid = false;
				}
			});
			if (valid === true) {
				this.endMessages = msg;
			}
		}
	}
}


src/extension.ts
import * as vscode from "vscode";
import { Timer } from "./Timer";

export function activate({ subscriptions }: vscode.ExtensionContext) {
	const config = vscode.workspace.getConfiguration("BreakTime"); //user configuration

	if (config.enable === true) {
		const timerStatusBarItem: vscode.StatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 299);
		const timerStatusBarTxt: string = `Next Break`;

		//Timer
		const timer = new Timer(timerStatusBarItem, timerStatusBarTxt);
		const timerCommandId: string = "BreakTime.start";

		subscriptions.push(
			vscode.commands.registerCommand(timerCommandId, () => {
				//Initialize quickpick
				const quickPick = vscode.window.createQuickPick();

				const quickPickItems: vscode.QuickPickItem[] = [
					{ label: "1 min" },
					{ label: "5 min" },
					{ label: "10 min" },
					{ label: "20 min" },
					{ label: "25 min" },
					{ label: "30 min" },
					{ label: "40 min" },
					{ label: "60 min" },
					{ label: "90 min" },
					{ label: "120 min" },
				];
				//if the timer is running add an item to stop it
				const stopLabel = "STOP";
				if (timer.isRunning()) {
					quickPickItems.unshift({ label: stopLabel, description: "Stop the timer" });
				}

				quickPick.items = quickPickItems;

				//user picked
				quickPick.onDidChangeSelection(([item]) => {
					if (item) {
						//set the time picked and start the timer
						let time = Number(item.label.split(" ")[0]) * 60;
						if (item.label === stopLabel) {
							time = 0;
						}
						timer.start(time);
					}
					quickPick.hide();
				});
				quickPick.show();
			}),
		);

		timerStatusBarItem.command = timerCommandId;
		timerStatusBarItem.text = timerStatusBarTxt;
		timerStatusBarItem.show();
		subscriptions.push(timerStatusBarItem);
	}
}

// this method is called when your extension is deactivated
export function deactivate() {}


out/extension.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.deactivate = exports.activate = void 0;
const vscode = require("vscode");
const Timer_1 = require("./Timer");
function activate({ subscriptions }) {
    const config = vscode.workspace.getConfiguration("BreakTime"); //user configuration
    if (config.enabled === true) {
        const timerStatusBarItem = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 299);
        const timerStatusBarTxt = `Next Break`;
        //Timer
        const timer = new Timer_1.Timer(timerStatusBarItem, timerStatusBarTxt);
        const timerCommandId = "BreackTime.start";
        subscriptions.push(vscode.commands.registerCommand(timerCommandId, () => {
            //Initialize quickpick
            const quickPick = vscode.window.createQuickPick();
            const quickPickItems = [
                { label: "1 min" },
                { label: "5 min" },
                { label: "10 min" },
                { label: "20 min" },
                { label: "25 min" },
                { label: "30 min" },
                { label: "40 min" },
                { label: "60 min" },
                { label: "90 min" },
                { label: "120 min" },
            ];
            //if the timer is running add an item to stop it
            const stopLabel = "STOP";
            if (timer.isRunning()) {
                quickPickItems.unshift({ label: stopLabel, description: "Stop the timer" });
            }
            quickPick.items = quickPickItems;
            //user picked
            quickPick.onDidChangeSelection(([item]) => {
                if (item) {
                    //set the time picked and start the timer
                    let time = Number(item.label.split(" ")[0]) * 60;
                    if (item.label === stopLabel) {
                        time = 0;
                    }
                    timer.start(time);
                }
                quickPick.hide();
            });
            quickPick.show();
        }));
        timerStatusBarItem.command = timerCommandId;
        timerStatusBarItem.text = timerStatusBarTxt;
        timerStatusBarItem.show();
        subscriptions.push(timerStatusBarItem);
    }
}
exports.activate = activate;
// this method is called when your extension is deactivated
function deactivate() { }
exports.deactivate = deactivate;
//# sourceMappingURL=extension.js.map